### PR TITLE
aarch64: Use asid_map_t bitfield to store vspace attributes to free mapping slots

### PR DESCRIPTION
**Pull request by kent-mcleod**
Use the asid_map bitfield from x86 for aarch64 to map from ASID to an intermediate asid_map object that can hold a vspace object reference plus additional related attributes.  This can free up reserved slots in the top level page table for use at user level. 

This PR uses some commits from https://github.com/seL4/seL4/pull/579 but is only targeting aarch64.

More background: https://sel4.atlassian.net/browse/SELFOUR-1840

### CODE REVIEW

File: include/arch/arm/arch/64/mode/object/structures.bf
Comment (lsf37) [inline include/arch/arm/arch/64/mode/object/structures.bf:69]: This commit also changes the width of the CB field in the caps. Was that intentional? If yes, should probably be mentioned in the commit message. I haven't checked all other `cb` occurrences if the new width fits or may lead to overflow.

File: src/fastpath/fastpath.c
Comment (lsf37) [inline src/fastpath/fastpath.c:112]:
```suggestion
    /* Ensure the vmid is valid. */
    if (unlikely(!asid_map_asid_map_vspace_get_stored_vmid_valid(asid_map))) {
        slowpath(SysCall);
```

Comment (lsf37) [reply src/fastpath/fastpath.c:112]: Given the difference in naming in the other configs (`hw_asid`), might need a comment that `vmid`s are the tags used in hyp mode.

File: src/fastpath/fastpath.c
Comment (lsf37) [inline src/fastpath/fastpath.c:379]:
```suggestion
    /* Ensure the vmid is valid. */
```

File: src/arch/arm/64/kernel/vspace.c
Comment (lsf37) [inline src/arch/arm/64/kernel/vspace.c:574]:
```suggestion
    asid_pool_t *poolPtr;
```

(at least as far as I can see in the diff, there are no other declarations to align with any more)

File: src/arch/arm/64/kernel/vspace.c
Comment (lsf37) [inline src/arch/arm/64/kernel/vspace.c:1086]: is the additional indent intentional?

File: src/arch/arm/64/kernel/vspace.c
Comment (lsf37) [inline src/arch/arm/64/kernel/vspace.c:1084]: There is is this convention that the functions that expect things to maybe not be there are called `find` and return a potential failure. This one is more a `get`, because it asserts the presence of the ASID pool. I currently have the same `find` name in the prelim spec, based on Curtis' original commit and probably going back to x64, and it makes sense to mimic `findMapForASID`, but overall I think `getMapRefForASID` would be a better name.

Comment (lsf37) [reply src/arch/arm/64/kernel/vspace.c:1084]: Actually, I lie, I don't have this function at all in the spec -- I need to unfold this in Haskell because there is no direct ref to a position in an array there. Anyway, still prefer the `get` name :-)

File: src/arch/arm/64/kernel/vspace.c
Comment (lsf37) [inline src/arch/arm/64/kernel/vspace.c:1203]:
```suggestion
```

File: src/arch/arm/64/kernel/vspace.c
Comment (lsf37) [inline src/arch/arm/64/kernel/vspace.c:1204]:
```suggestion
    asid_map_asid_map_vspace_ptr_set_bind_cb(asid_map, asid_map_asid_map_vspace_ptr_get_bind_cb(asid_map) + 1);
}
```

File: src/arch/arm/64/kernel/vspace.c
Comment (kent-mcleod) [reply src/arch/arm/64/kernel/vspace.c:1084]: Ok, get makes more sense.

File: src/arch/arm/64/kernel/vspace.c
Comment (kent-mcleod) [reply src/arch/arm/64/kernel/vspace.c:1086]: It came in from the cherry-pick. I assumed the style tool would fix it but it seems not.

File: src/fastpath/fastpath.c
Comment (kent-mcleod) [reply src/fastpath/fastpath.c:112]: I went with the vmid naming because of the discussion on the draft PR, but to be more complete, all the other function names should probably be changed too...

File: src/arch/arm/64/kernel/vspace.c
Comment (lsf37) [inline src/arch/arm/64/kernel/vspace.c:2409]: If for some reason the previous shorter version is preferred, I think it would be Ok to leave it, because it should be equivalent. My guess is that `-O2` will reduce it to the same assembly anyway, though, and this is slightly clearer.

File: src/fastpath/fastpath.c
Comment (lsf37) [reply src/fastpath/fastpath.c:112]: Yes, I agree, was just about to write that in a summary comment. Might make sense do that rename in a separate commit/PR.

**Review by lsf37 — APPROVED** [review]
Very nice, a whole lot of stuff is becoming clearer and most parts shorter.

File: src/fastpath/fastpath.c
Comment (lsf37) [reply src/fastpath/fastpath.c:112]: Probably even the type `hw_asid_t`, if that doesn't conflict with code that is shared between configs (might be a problem in the fastpath for instance).

**Comment by lsf37** [comment]
Cool, I very much like this. Would be good to have another opinion from a systems person, maybe @yyshen if you have time?

File: include/arch/arm/arch/64/mode/object/structures.bf
Comment (yyshen) [inline include/arch/arm/arch/64/mode/object/structures.bf:258]: recent processors support 16-bit vmid, so we could just make it 16b-bit.

File: include/arch/arm/arch/64/mode/object/structures.bf
Comment (kent-mcleod) [reply include/arch/arm/arch/64/mode/object/structures.bf:69]: Intentional. According to the other PR and the SMMUv2 spec, there is a maximum number of context banks of 128 which requires 8 bits to include a valid bit. I tried to update the width in all of the other locations.  The asid_map_t is short of available bit space and I saw that there's already a push for changing them to 8bits in other PRs so figured to make this part of the change at the same time.

File: src/arch/arm/64/kernel/vspace.c
Comment (kent-mcleod) [reply src/arch/arm/64/kernel/vspace.c:574]: Yea, it came in with a cherry-picked commit and I assumed the style tool would fix it up.....

File: include/arch/arm/arch/64/mode/object/structures.bf
Comment (kent-mcleod) [reply include/arch/arm/arch/64/mode/object/structures.bf:258]: On a processor that supports 16bit VMID we would be able to just use the asid_t value which is also 16bits and not need a stored_hw_vmid slot.

**Comment by yyshen** [comment]
> Cool, I very much like this. Would be good to have another opinion from a systems person, maybe @yyshen if you have time?

@lsf37 How did you do the review based on github? Sometimes I found it is hard to understand the changes by looking at the diffs on Github. Any suggestions?

File: include/arch/arm/arch/64/mode/object/structures.bf
Comment (yyshen) [reply include/arch/arm/arch/64/mode/object/structures.bf:258]: right, if we have enough hardware asid, we do not need the mapping. where do we specify that?

**Comment by kent-mcleod** [comment]
> @lsf37 How did you do the review based on github? Sometimes I found it is hard to understand the changes by looking at the diffs on Github. Any suggestions?

Do you browse  the diffs commit by commit? I think that's the clearest you can get using the web client. If that's not good enough for me I manually clone and step through diffs locally.  Onetime I tried using GitHub's desktop app which was a bit better for quickly stepping through files and commits for a PR but you need more setup...

**Comment by lsf37** [comment]
> Do you browse the diffs commit by commit? I think that's the clearest you can get using the web client. If that's not good enough for me I manually clone and step through diffs locally. Onetime I tried using GitHub's desktop app which was a bit better for quickly stepping through files and commits for a PR but you need more setup...

Yeah, I do all of these for large PRs, although the hurdle is pretty high for the GH desktop app. I find the diff view in VSCode pretty good (not always, but on average). For this specific one, I got by with looking at the single commits and long staring at the diff. It helps that I wrote a spec for this separately, so it was easy for me to recognise patterns.

**Comment by yyshen** [comment]
> > Do you browse the diffs commit by commit? I think that's the clearest you can get using the web client. If that's not good enough for me I manually clone and step through diffs locally. Onetime I tried using GitHub's desktop app which was a bit better for quickly stepping through files and commits for a PR but you need more setup...
> 
> Yeah, I do all of these for large PRs, although the hurdle is pretty high for the GH desktop app. I find the diff view in VSCode pretty good (not always, but on average). For this specific one, I got by with looking at the single commits and long staring at the diff. It helps that I wrote a spec for this separately, so it was easy for me to recognise patterns.



> > Do you browse the diffs commit by commit? I think that's the clearest you can get using the web client. If that's not good enough for me I manually clone and step through diffs locally. Onetime I tried using GitHub's desktop app which was a bit better for quickly stepping through files and commits for a PR but you need more setup...
> 
> Yeah, I do all of these for large PRs, although the hurdle is pretty high for the GH desktop app. I find the diff view in VSCode pretty good (not always, but on average). For this specific one, I got by with looking at the single commits and long staring at the diff. It helps that I wrote a spec for this separately, so it was easy for me to recognise patterns.

Thanks for the tips. I found the split view works better for me.

File: src/arch/arm/64/kernel/vspace.c
Comment (lsf37) [reply src/arch/arm/64/kernel/vspace.c:2409]: Yes, you're right, that doesn't really work. I tend to ignore C's type system because it doesn't guarantee anything :-) I was only looking at the representation, which hasn't changed for the False case.

File: include/arch/arm/arch/64/mode/object/structures.bf
Comment (lsf37) [reply include/arch/arm/arch/64/mode/object/structures.bf:69]: Ok, that makes sense. Would be good to mention in the commit message. I have checked the other occurrences, and at least the types are safe everywhere, so this is good from my side.

File: src/arch/arm/64/kernel/vspace.c
Comment (yyshen) [inline src/arch/arm/64/kernel/vspace.c:1407]: additional spaces?

File: include/arch/arm/arch/64/mode/kernel/vspace.h
Comment (yyshen) [inline include/arch/arm/arch/64/mode/kernel/vspace.h]: the change is mostly for arm64 hyp right? for the cores with 8-bit VMID. So the commit message could point out the VMID explicitly I guess.

File: src/fastpath/fastpath.c
Comment (yyshen) [inline src/fastpath/fastpath.c:104]: do we know how many cycles are added by this findMapForASID in fastpath?

File: include/arch/arm/arch/64/mode/object/structures.bf
Comment (yyshen) [reply include/arch/arm/arch/64/mode/object/structures.bf:69]: CONFIG_ARM_SMMU seems implying v2. We probably will need to support V3 soon.

**Review by yyshen — APPROVED** [review]
looks good to me. Thanks!

File: src/arch/arm/64/kernel/vspace.c
Comment (lsf37) [reply src/arch/arm/64/kernel/vspace.c:1407]: it's now under an additional `if` so the new indent looks correct to me

File: src/arch/arm/64/kernel/vspace.c
Comment (lsf37) [reply src/arch/arm/64/kernel/vspace.c:1202]: Good point. This is independent of this PR, but I think there is a bug in this path, yes, because the decode path does not check that the increment is safe (if assertions are turned on, this would fail a bf generator assertion I think; if they are turned off it would wrap, which is not clear that we want here).

The check should be in decode, though, not here, because in here it is too late to do anything about it. It's actually unclear to me how this is supposed to work when the CB space is used up.

We should create an issue for this.

Comment (kent-mcleod) [reply src/arch/arm/64/kernel/vspace.c:1202]: There is a maximum number of CBs allowed in an SMMU implementation defined by the spec to be 128 and CBs can only be assigned to a single VSpace. 8 bits was chosen to be large enough to support every CB being assigned to the same VSpace on any SMMUv2 implementation so it shouldn't be able to overflow due to existing checks in `decodeARMCBInvocation` and `decodeARMCBControlInvocation`.

File: src/fastpath/fastpath.c
Comment (kent-mcleod) [reply src/fastpath/fastpath.c:104]: There's a few changes here that affect the fastpath, so I'll try and break them down.

### Effect of asid_map_t and extra asid_t check on aarch64 non-hyp

aarch64 non-hyp was missing a check for the ASID validity. This means that if an asid_pool_t containing the target thread's VSpace was revoked and then a fastpath was taken the VSpace of the target thread would be switched to even though the VSpace would not have a valid ASID.  The correct behavior would be for an empty table to be assigned to the TTBR as part of the context switch which is what falling back to the slowpath achieves.

For this change I measured a 1 cycle increase in the published numbers (https://sel4.systems/About/Performance/) for default IPC call and reply on the TX1 and TX2. The program should be doing 2 extra memory reads and 2 extra comparisons. I didn't check any event counters yet though and am assuming that the CPU is just making the right magical optimizations and this isn't adding any new stalls.

Any cost increase should be acceptable here as it's adding a check that should be required.

### Effect of asid_map_t and extra asid_t check on aarch64 hyp

aarch64 non-hyp was also missing an asid validity check and would finally call `getHWASID` as part of the context switch. If the asid wasn't valid here then the kernel may dereference a null pointer. As part of these changes, `getHWASID` was removed from the fastpath and replacing it with the changes in the PR the fastpath ends up a few % faster.  I don't have the numbers written down, but can update before final merge. 

### Effect of asid_map_t and removing ASID from page table entry on aarch32 sabre

These changes aren't part of this PR, but I spent some time evaluating an aarch32 implementation assuming it would be most sensitive to costs of this change.  On the sabre, a memory look up is removed as the hw_asid is taken out of the reserved page table entry, but 2 lookups and a conditional branch are added for the ASID table lookups. Published numbers go from 317 cycles (call)  and 336 cycles (recv) on Sabre to 333 and 350 now.  So 16 and 14 more cycles which is 4-5% overhead.

File: include/arch/arm/arch/64/mode/kernel/vspace.h
Comment (kent-mcleod) [reply include/arch/arm/arch/64/mode/kernel/vspace.h]: The change is for aarch64 hyp and also SMMUv2 and eventually kernel image clone. Longer term there is a plan that every architecture will have the asid_map_t feature if it wants to support the kernel image clone feature (a vspace can have a different kernel image to avoid shared cache timings).

File: include/arch/arm/arch/64/mode/object/structures.bf
Comment (kent-mcleod) [reply include/arch/arm/arch/64/mode/object/structures.bf:258]: At the moment all of the supported platforms only have 8bit VMID. If there's a platform that supports 16bit VMID in the future then the kernel would get a config update to make this feature `CONFIG_ARM_HYPERVISOR_SUPPORT && CONFIG_8BIT_VMID`

File: include/arch/arm/arch/64/mode/kernel/vspace.h
Comment (axel-h) [inline include/arch/arm/arch/64/mode/kernel/vspace.h:73]: nitpicking: go for c99?
```suggestion
    asid_map_t asid_map = asid_map_asid_map_vspace_new(
```

File: include/arch/arm/arch/64/mode/kernel/vspace.h
Comment (axel-h) [inline include/arch/arm/arch/64/mode/kernel/vspace.h:120]: nitpicking: go for c99?
```suggestion
    asid_map_t asid_map = asid_map_asid_map_vspace_new(
```

File: libsel4/sel4_arch_include/aarch64/sel4/sel4_arch/constants.h
Comment (axel-h) [inline libsel4/sel4_arch_include/aarch64/sel4/sel4_arch/constants.h:268]: Could go for a more intuitive 
```suggestion
#define seL4_UserTop ((1 << 40) - 1)
```
now. Not sure of `(LIBSEL4_BIT(40) - 1)` also work, as this enforced 'ul' then.

Comment (kent-mcleod) [reply libsel4/sel4_arch_include/aarch64/sel4/sel4_arch/constants.h:268]: Using a plain old constant is what all other definitions use.

Comment (lsf37) [reply libsel4/sel4_arch_include/aarch64/sel4/sel4_arch/constants.h:268]: Would agree that it makes more sense to stay consistent with the rest here, although I do generally appreciate the form that tells me where the constant is coming from. It's reasonably easy to see here, though, so not a big deal.

File: src/arch/arm/64/kernel/vspace.c
Comment (lsf37) [reply src/arch/arm/64/kernel/vspace.c:1202]: In any case it is a separate issue, happy to leave it for another PR or an issue (still need to double check that decode path -- the first time I looked I didn't see a check that would constrain the overflow)

Comment (kent-mcleod) [reply src/arch/arm/64/kernel/vspace.c:1202]: It's constrained by the upper bound on number of CBs that can exist in the system which is the number of slots in the static CB table: `extern bool_t smmuStateCBTable[SMMU_MAX_CB];`  SMMU_MAX_CB cannot be greater than 128.

**Comment by kent-mcleod** [comment]
Just waiting for this to pass hw tests before merging

**Pull request merged**
